### PR TITLE
Item Flags Support

### PR DESCRIPTION
**Pull request by EquipableMC**
### Description
This PR adds support for ItemFlags. This is taken directly from SkBee but modified to support Skript. All credit goes to @ShaneBeee 

---
**Target Minecraft Versions:** any
**Requirements:** none
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprItemFlags.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemFlag;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+@Name("ItemFlag - The ItemFlags of an Item")
```
Comment (Asleeepp) [inline src/main/java/ch/njol/skript/expressions/ExprItemFlags.java:23]: Change the name, Skript doesn't follow this naming scheme

File: src/main/java/ch/njol/skript/expressions/ExprItemFlags.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	@SuppressWarnings("NullableProblems")
+	@Override
+	public Class<?> @Nullable [] acceptChange(Changer.ChangeMode mode) {
+		if (mode == Changer.ChangeMode.ADD || mode == Changer.ChangeMode.REMOVE) {
+			return CollectionUtils.array(ItemFlag[].class);
+		} else if (mode == Changer.ChangeMode.DELETE || mode == Changer.ChangeMode.RESET) {
+			return CollectionUtils.array();
+		}
+		return null;
+	}
```
Comment (Asleeepp) [inline src/main/java/ch/njol/skript/expressions/ExprItemFlags.java:64]: use a switch

File: src/main/java/ch/njol/skript/expressions/ExprItemFlagsItem.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.bukkit.inventory.ItemFlag;
+import org.bukkit.inventory.meta.ItemMeta;
+import org.jetbrains.annotations.Nullable;
+
+@Name("ItemFlag - Item with ItemFlags")
```
Comment (Asleeepp) [inline src/main/java/ch/njol/skript/expressions/ExprItemFlagsItem.java:19]: same here

**Review by Asleeepp — CHANGES_REQUESTED** [review]

**Comment by ShaneBeee** [comment]
As discussed on discord.
Please learn java and write this yourself.
I personally do not appreciate you copy/pasting my code and taking credit for it.
And no, pasting my username in the body of a PR is NOT crediting anyone.

**Comment by sovdeeth** [comment]
I think these syntaxes should be rewritten from scratch. I'm not very comfortable with the precedent of copying from an addon and slightly adapting it for Skript.

**Comment by EquipableMC** [comment]
> I think these syntaxes should be rewritten from scratch. I'm not very comfortable with the precedent of copying from an addon and slightly adapting it for Skript.

dont know how to rewrite it from scratch then as I know I'll just make it look very similar

**Comment by sovdeeth** [comment]
> > I think these syntaxes should be rewritten from scratch. I'm not very comfortable with the precedent of copying from an addon and slightly adapting it for Skript.
> 
> dont know how to rewrite it from scratch then as I know I'll just make it look very similar

That's ok, there's only so many ways to write something. What's not ok is copy-pasting the code, without credit. If I were to write this I'd also end up with something pretty similar, but that's due to structure and design choices, not due to copy-pasting.

**Comment by EquipableMC** [comment]
> > > I think these syntaxes should be rewritten from scratch. I'm not very comfortable with the precedent of copying from an addon and slightly adapting it for Skript.
> > 
> > 
> > dont know how to rewrite it from scratch then as I know I'll just make it look very similar
> 
> That's ok, there's only so many ways to write something. What's not ok is copy-pasting the code, without credit. If I were to write this I'd also end up with something pretty similar, but that's due to structure and design choices, not due to copy-pasting.

Again, I don't know how else to rewrite it 🤷‍♂️

**Comment by sovdeeth** [comment]
I'm not very concerned with exactly what shane said, I'm concerned with the precedent this sets for future PRs. My concerns center around people getting anchored on one way to do something (the way the addon did it) without adequately exploring other options, designs, or methods. It also leads to code in the code base that contributors aren't as familiar with compared to as if one wrote it themselves. If you don't think you could write this on your own, my personal opinion is that you should leave it for someone else. 

Taking inspiration from addon implementations is fine and dandy and often beneficial, but the process of writing up the syntaxes, writing tests for them, and trying them out yourself is valuable in revealing other paths and improvements you could make, as well as finding possible bugs.

**Pull request closed without merging**